### PR TITLE
[4/N] Add test for partial training for NamedOptimizer

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91344

### CODE REVIEW

File: torch/distributed/optim/named_optimizer.py [hunk 1]
```diff
@@ -248,10 +239,9 @@ def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
                 param_keys.append(self.ordered_param_keys[param_key])  # type: ignore[call-overload]
             new_group_map[_gen_param_group_key(param_keys)] = new_group
         for group_key, new_group in new_group_map.items():
+            # When the conditional training is performed, not all parameters need to be in the param_group.
```
Comment (rohan-varma) [inline torch/distributed/optim/named_optimizer.py:242]: I'd add some more detail here as "conditional training" is not really a well-known term IMO, maybe something like

"When not all parameters are used in training or receive gradient, not all parameters would be in the `param_group`. Thus we <whatever logic happens in this case>".

File: test/distributed/optim/test_named_optimizer.py [hunk 2]
```diff
@@ -291,6 +291,56 @@ def test_load_state_dict(self):
...
+            torch.optim.SGD,
+            [
+                {"params": m.net1.parameters()},
+                {"params": m.net3.parameters(), "lr": 1e-3},
+            ],
+            lr=1e-2,
+            momentum=0.9,
+        )
+
+        for _ in range(2):
```
Comment (rohan-varma) [inline test/distributed/optim/test_named_optimizer.py:308]: sorry I may miss something, but where is conditional training happening here?

**Review by rohan-varma — APPROVED** [review]

File: test/distributed/optim/test_named_optimizer.py (hunk 2 above)
Comment (fduwjj) [reply test/distributed/optim/test_named_optimizer.py:308]: So I only set parameters from net1 and net3 into param groups for the optim_1, while for optim_2, it is updating all parameters. So, the loading from the state from optim_1 to optim_2 only changes for the corresponding parameters.

File: torch/distributed/optim/named_optimizer.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/optim/named_optimizer.py:242]: Sure, thanks for the suggestion.
